lab_4/main.py

```diff
@@ -3,45 +3,118 @@
-    pass
+    if not isinstance(text, str):
+        raise ValueError
+
+    sentences = re.split(r'[.!?]', text)
+    res = []
+
+    for sentence in sentences:
+        clean_sentence = re.sub('[^a-z \n]', '', sentence.lower()).split()
+        if clean_sentence:
+            res.extend(clean_sentence + ['<END>'])
+
+    return tuple(res)
 
 
 class WordStorage:
     def __init__(self):
-        pass
+        self.storage = {}
 
     def _put_word(self, word: str):
-        pass
+        if not isinstance(word, str) or not word:
+            raise ValueError
+        if word not in self.storage:
+            self.storage[word] = len(self.storage) + 1
+        return self.storage[word]
 
     def get_id(self, word: str) -> int:
-        pass
+        if not isinstance(word, str) or not word:
+            raise ValueError
+        if word not in self.storage:
+            raise KeyError
+        return self.storage[word]
 
     def get_word(self, word_id: int) -> str:
-        pass
+        if not isinstance(word_id, int):
+            raise ValueError
+        if word_id not in self.storage.values():
+            raise KeyError
+        for word in self.storage:
+            if self.storage[word] == word_id:
+                return word
 
     def update(self, corpus: tuple):
-        pass
+        if not isinstance(corpus, tuple):
+            raise ValueError
+        for word in corpus:
+            self._put_word(word)
+        return 0
 
 
 def encode_text(storage: WordStorage, text: tuple) -> tuple:
-    pass
+    if not isinstance(storage, WordStorage) or not isinstance(text, tuple):
+        raise ValueError
+
+    return tuple([storage.get_id(word) for word in text])
 
 
 class NGramTextGenerator:
     def __init__(self, word_storage: WordStorage, n_gram_trie: NGramTrie):
-        pass
+        self._word_storage = word_storage
+        self._n_gram_trie = n_gram_trie
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+        if not isinstance(context, tuple) or len(context) != self._n_gram_trie.size - 1:
+            raise ValueError
+
+        context_n_grams = {}
+        for n_gram, freq in self._n_gram_trie.n_gram_frequencies.items():
+            if context == n_gram[:self._n_gram_trie.size - 1]:
+                context_n_grams[n_gram] = freq
+
+        if not context_n_grams:
+            finding_top_word = sorted(self._n_gram_trie.uni_grams, key=self._n_gram_trie.uni_grams.get, reverse=True)
+            return finding_top_word[0][0]
+
+        finding_top_context_n_gram = sorted(context_n_grams, key=context_n_grams.get, reverse=True)
+        return finding_top_context_n_gram[0][-1]
 
     def _generate_sentence(self, context: tuple) -> tuple:
-        pass
+        if not isinstance(context, tuple):
+            raise ValueError
+
+        new_context = []
+        new_context.extend(list(context))
+        counter = 0
+        while counter < 20:
+            counter += 1
```

Comment:
If there is a finite number of times you are going to run through a loop, it is better to use another type of loop.